src/driver/amdxdna/ve2_debug.c

```diff
@@ -0,0 +1,359 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2025, Advanced Micro Devices, Inc.
+ */
+#include <linux/device.h>
+#include <linux/version.h>
+
+#include "ve2_fw.h"
+#include "ve2_mgmt.h"
+#include "ve2_of.h"
+
+static int ve2_get_hwctx_status(struct amdxdna_client *client, struct amdxdna_drm_get_info *args)
+{
+	size_t hwctx_data_sz = sizeof(struct amdxdna_drm_query_ctx);
+	struct amdxdna_drm_query_ctx *hwctx_data;
+	struct amdxdna_drm_query_ctx __user *buf;
+	struct amdxdna_dev *xdna = client->xdna;
+	struct amdxdna_client *tmp_client;
+	u32 req_bytes = 0, hw_i = 0;
+	struct amdxdna_ctx *hwctx;
+	unsigned long hwctx_id;
+	bool overflow = false;
+	int ret = 0, idx;
+
+	hwctx_data = kzalloc(hwctx_data_sz, GFP_KERNEL);
+	if (!hwctx_data)
+		return -ENOMEM;
+
+	buf = u64_to_user_ptr(args->buffer);
+
+	list_for_each_entry(tmp_client, &xdna->client_list, node) {
+		idx = srcu_read_lock(&tmp_client->ctx_srcu);
+		amdxdna_for_each_ctx(tmp_client, hwctx_id, hwctx) {
+			req_bytes += hwctx_data_sz;
+			if (args->buffer_size < req_bytes) {
+				/* Continue iterating to get the required size */
+				overflow = true;
+				continue;
+			}
+
+			hwctx_data->pid = hwctx->client->pid;
+			hwctx_data->context_id = hwctx->id;
+			hwctx_data->start_col = hwctx->start_col;
+			hwctx_data->num_col = hwctx->num_col;
+			XDNA_DBG(xdna, "cl_pid: %llu, hwctx_id: %u, start_col %u, ncol %u\n",
+				 hwctx_data->pid, hwctx_data->context_id, hwctx_data->start_col,
+				 hwctx_data->num_col);
+
+			if (copy_to_user(&buf[hw_i], hwctx_data, hwctx_data_sz)) {
+				ret = -EFAULT;
+				srcu_read_unlock(&tmp_client->ctx_srcu, idx);
+				goto out;
+			}
+			hw_i++;
+		}
+		srcu_read_unlock(&tmp_client->ctx_srcu, idx);
+	}
+
+	if (overflow) {
+		XDNA_ERR(xdna, "Invalid buffer size. Given: %u Need: %u", args->buffer_size,
+			 req_bytes);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if (hw_i == 0) {
+		XDNA_ERR(xdna, "pid %d failed to get hwctx\n", client->pid);
+		ret = -EINVAL;
```

Comment:
Why does it return error here? Is it just args->buffer_size == 0 when there is not any hwctx?

Reply:
This above code block unable to find any active hwctx from the client_list, so return error status to caller.

Reply:
Updated the PR with this.